Replace `sympy.solve` with a new simplified one.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105877

This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

ysiraichi committed:
Replace `sympy.solve` with a new simplified one.

This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Fix solution not found. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -3183,7 +3184,12 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
+                r = try_solve(expr, free[0])
+                if r is None or r[1] != solution:
+                    rstr = "None" if r is None else r[1]
+                    print(f"EQ ({free[0]}): {lhs} = {rhs} ==> {solution} != {rstr} (ours)")
```
Left these prints so that I can see cases where `sympy.solve` works but our doesn't.

ysiraichi committed:
Add multiplication rule. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Fix simplification rules. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Only refine range of integer symbols. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,120 @@
+#   2. The expression on the right-hand side
+# Returns 'None' if it can't reach a state where the only thing in the left
+# hand side is 'thing'.
+def try_solve(
```
I'd suggest thinking if there's some good testing strategy for this function directly.

ysiraichi replied:
I guess we could, at least, try to cover all execution paths. Do you think that would be enough?
Maybe we could also take a few examples that appear when running some models. What do you think?

lezcano replied:
Could we have a list of inputs that exercises one and only one of the paths and then prove with Z3 that the result is equivalent?

ezyang commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,120 @@
+        # a // b <= expr => a < b * (expr + 1)
+        if isinstance(expr, (sympy.Lt, sympy.Le)):
+            quotient = rhs if isinstance(expr, sympy.Lt) else (rhs + 1)  # type: ignore[arg-type]
+            return sympy.Lt(lhs.args[0], (quotient * lhs.args[1]))  # type: ignore[arg-type]
```
You have fairly similar logic at https://github.com/pytorch/pytorch/blob/d54cf0254328a56bb5615465e7f5ea5e1799d1cd/torch/fx/experimental/symbolic_shapes.py#L3010-L3026

Is the duplication intentional?

ysiraichi replied:
No. Forgot to delete that.

ysiraichi committed:
Add tests. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi committed:
Fix lint issues. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,167 @@
+        assert isinstance(e, sympy.Rel)
+
+        if not e.lhs.has(thing):
+            continue
```
There's probably some micro-optimizations here (avoiding repeated .has() traversals) which you could apply if so inclined.

ezyang commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,167 @@
+
+        # Divide both sides by 'other'.
+        lhs = lhs / other
+        rhs = rhs / other
```
push the division into the conditional?

ezyang reviewed (approved):
Nice refactor, thanks!

ysiraichi committed:
Update on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

nunoplopes commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,175 @@
+    # Give up when 'thing' appears on both sides of the relational expression.
+    # That is because, as is, we assume the thing we are trying to isolate is
+    # only on the right-hand side.
+    if lhs_has_thing and rhs_has_thing:
```
this skips optimization of smth like `2*thing >= thing + 1` → `thing >= 1`. Is that intended?

ysiraichi replied:
Yes. At this point, the simplification rules only look at the left-hand side, trying to isolate `thing`. Once we implement more rules, I guess we could remove this limitation. But, as far as I've see, this is enough for the benchmarks we have. Meaning that, it yields the same solutions comparing to `sympy.solve` for the places it was being used.

nunoplopes commented on torch/utils/_sympy/solve.py:
```diff
@@ -0,0 +1,175 @@
+    # Given the expression: a // b op c
+    # where 'op' is a relational operation, these rules only work if:
+    #   - b > 0
+    #   - c is an integer
```
this is all correct according to Z3:
```smt
(declare-const a Real)
(declare-const b Real)
(declare-const c Int)

(assert (> b 0))

(push)
(assert (not (=
  (= (to_int (/ a b)) c)
  (and (>= a (* b c)) (< a (* b (+ c 1))))
)))
(check-sat)
(pop)

(push)
(assert (not (=
  (not (= (to_int (/ a b)) c))
  (or (< a (* b c)) (>= a (* b (+ c 1))))
)))
(check-sat)
(pop)

(push)
(assert (not (=
  (> (to_int (/ a b)) c)
  (>= a (* b (+ c 1)))
)))
(check-sat)
(pop)

(push)
(assert (not (=
  (>= (to_int (/ a b)) c)
  (>= a (* b (+ c 0)))
)))
(check-sat)
(pop)

(push)
(assert (not (=
  (< (to_int (/ a b)) c)
  (< a (* b (+ c 0)))
)))
(check-sat)
(pop)

(push)
(assert (not (=
  (<= (to_int (/ a b)) c)
  (< a (* b (+ c 1)))
)))
(check-sat)
(pop)
```

ysiraichi commented:
@pytorchbot merge

ysiraichi committed:
Rebased. on "Replace `sympy.solve` with a new simplified one."


This PR implements `try_solve`: a function that tries to move terms of a relational
expression around, so as to isolate a given variable on the left-hand side.

For example:

```python
>>> try_solve(Eq(a + 5, 3), a)
Eq(a, -2)
>>> try_solve(Gt(Mod(a, 3), 0), a) # returns None
>>> try_solve(Gt(Mod(a, 3), 0), Mod(a, 3))
Gt(Mod(a, 3), 0), Mod(a, 3)
```

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge
